hail/src/main/scala/is/hail/expr/ir/ForwardLets.scala

```diff
@@ -0,0 +1,97 @@
+package is.hail.expr.ir
+
+import is.hail.utils._
+
+import scala.collection.mutable
+
+object ForwardLets {
+  def apply(ir0: BaseIR): BaseIR = {
+    val ir1 = new NormalizeNames(allowFreeVariables = true).apply(ir0)
+    val UsesAndDefs(uses, _) = ComputeUsesAndDefs(ir1, freeVariablesError = false)
+    val nestingDepth = NestingDepth(ir1)
+
+    def rewriteTable(tir: TableIR): BaseIR = tir.copy(tir
+      .children
+      .iterator
+      .zipWithIndex
+      .map {
+      case (child: IR, i) => rewrite(child, NewBindings(tir, i).dropBindings)
+      case (tir: TableIR, _) => rewriteTable(tir)
+      case (mir: MatrixIR, _) => rewriteMatrix(mir)
+      case (bmir: BlockMatrixIR, _) => rewriteBlockMatrix(bmir)
+    }.toFastIndexedSeq)
+
+    def rewriteMatrix(mir: MatrixIR): BaseIR = mir.copy(mir
+      .children
+      .iterator
+      .zipWithIndex
+      .map {
+        case (child: IR, i) => rewrite(child, NewBindings(mir, i).dropBindings)
+        case (tir: TableIR, _) => rewriteTable(tir)
+        case (mir: MatrixIR, _) => rewriteMatrix(mir)
+        case (bmir: BlockMatrixIR, _) => rewriteBlockMatrix(bmir)
+      }.toFastIndexedSeq)
+
+    def rewriteBlockMatrix(bmir: BlockMatrixIR): BaseIR = bmir.copy(bmir
+      .children
+      .iterator
+      .zipWithIndex
+      .map {
+        case (child: IR, i) => rewrite(child, NewBindings(bmir, i).dropBindings)
+        case (tir: TableIR, _) => rewriteTable(tir)
+        case (mir: MatrixIR, _) => rewriteMatrix(mir)
+        case (bmir: BlockMatrixIR, _) => rewriteBlockMatrix(bmir)
+      }.toFastIndexedSeq)
+
+    def rewrite(ir: IR, env: BindingEnv[IR]): IR = {
+
+      def shouldForward(value: IR, refs: mutable.Set[RefEquality[Ref]], base: IR): Boolean = {
+        value.isInstanceOf[Ref] ||
+          IsConstant(value) ||
+          refs.isEmpty ||
+          refs.size == 1 &&
+            nestingDepth.lookup(refs.head) == nestingDepth.lookup(base) &&
+            !ContainsScan(value) &&
+            !ContainsAgg(value)
```

Comment:
Why can't we forward aggs and scans?  It should be OK if we're at the same nesting depth, no?

Reply:
The compiler does not properly handle mixed aggs and scans. This could easily forward an `ApplyScanOp` inside an `AggGroupBy` or something, and I hit errors in that case. We can relax this with some more testing and verification.

Reply:
Ah, yes, good point.  They are the same nesting depth but can't forward.  Thanks for clarifying!